storage/s3/options.go

```diff
@@ -44,26 +46,158 @@ var (
+	}
+	if parent.Region != "" {
+		ret.Region = &parent.Region
+	}
+	if parent.Key != "" && parent.Secret != "" {
+		ret.StaticCredentials = &StaticCredentials{
+			Key:    parent.Key,
+			Secret: parent.Secret,
+			Token:  parent.Token,
+		}
+	}
+	if parent.ExternalUri != "" {
+		ret.ExternalURI = &parent.ExternalUri
+	}
+	if parent.Uri != "" {
+		ret.URI = &parent.Uri
+	}
+	if parent.ForcePathStyle != ret.ForcePathStyle {
+		ret.ForcePathStyle = parent.ForcePathStyle
+	}
+	if parent.UseAccelerate != ret.UseAccelerate {
+		ret.UseAccelerate = parent.UseAccelerate
+	}
+	return ret
+}
+
+func (s storageSettings) Validate() error {
+	return validation.ValidateStruct(&s,
+		validation.Field(&s.StaticCredentials),
+	)
+}
+
+func (s storageSettings) options(opts *s3.Options) {
+	if s.StaticCredentials != nil {
+		opts.Credentials = *s.StaticCredentials
+	}
+	if s.Region != nil {
+		opts.Region = *s.Region
+	}
+	if s.URI != nil {
+		endpointURI := *s.URI
+		opts.EndpointResolver = s3.EndpointResolverFromURL(endpointURI,
+			func(ep *aws.Endpoint) {
+				ep.HostnameImmutable = s.ForcePathStyle
+			},
+		)
+	}
+	opts.UsePathStyle = s.ForcePathStyle
+	opts.UseAccelerate = s.UseAccelerate
+}
+
+func (s storageSettings) presignOptions(opts *s3.PresignOptions) {
+	if s.ExternalURI != nil {
+		presignURL := *s.ExternalURI
+		resolver := s3.EndpointResolverFromURL(presignURL,
+			func(ep *aws.Endpoint) {
+				ep.HostnameImmutable = s.ForcePathStyle
+			},
+		)
+		s3.WithPresignClientFromClientOptions(
+			s.options,
+			s3.WithEndpointResolver(resolver),
+		)(opts)
+	}
+}
+
+func (s *storageSettings) patch(setting *storageSettings) *storageSettings {
+	if setting == nil {
+		return s
+	}
+	if setting.BucketName != nil {
+		s.BucketName = setting.BucketName
+	}
+	if setting.StaticCredentials != nil {
+		s.StaticCredentials = setting.StaticCredentials
+	}
+	if setting.Region != nil {
+		s.Region = setting.Region
+	}
+	if setting.ExternalURI != nil {
+		s.ExternalURI = setting.ExternalURI
+	}
+	if setting.URI != nil {
+		s.URI = setting.URI
+	}
+	if setting.ProxyURI != nil {
+		s.ProxyURI = setting.ProxyURI
+	}
+	if setting.ForcePathStyle != s.ForcePathStyle {
+		s.ForcePathStyle = setting.ForcePathStyle
+	}
+	if setting.UseAccelerate != s.UseAccelerate {
+		s.UseAccelerate = setting.UseAccelerate
+	}
+	return s
+}
+
+func (s storageSettings) RewriteLink(link *model.Link) (*model.Link, error) {
+	if s.ProxyURI == nil {
+		return link, nil
+	}
+	uri, err := url.Parse(link.Uri)
+	if err != nil {
+		return link, err
+	}
+	new := s.ProxyURI.JoinPath(uri.Path)
+	q := new.Query()
+	for key, value := range uri.Query() {
+		q[key] = value
+	}
+	new.RawQuery = q.Encode()
+	new.Fragment = uri.Fragment
+	link.Uri = new.String()
+	return link, nil
+}
+
+type Options struct {
+	storageSettings
+
+	// ContentType of the uploaded objects
+	ContentType *string
+	// FilenameSuffix adds the suffix to the content-disposition for object downloads
+	FilenameSuffix *string
```

Comment:
not the easiest diff to read ;-)

Reply:
The first commit did a refactoring, segregating the `model.StorageSettings` part of `Options` into an internal `storageSettings` type. Maybe it's a bit overkill and I should just keep the (bigger) `model.StorageSettings` for readability although it contains some irrelevant settings?